README.md
# JADE-Cars MapSearching

## Name
assets - for creating maps, gps and screen (some sort of a simulation)

## Version
1.0.0

## Description
Codes and assets for search algorithms. The codes here are the refined and extended version of the original one. The codes here are also meant to suite the project based on the Eclipse IDE. For the stand-alone i.e. original codes, see [MapSearching](https://github.com/Ellednera/MapSearching) (the original codes were written and compilled manually)

Structure of the files:
 - ./assets/Station.java
 - ./assets/CityMap.java
 - ./assets/GPS.java
 - ./assets/Screen.java
 - ./TestScript.java - can be run independently, this is used to test the GUI and includes examples of how to use the algorithms
 - ./CSPBenchmark.java - benchmark for CSP
 - ./COPBenchmark.java - benchmark for COP

## Synopsis
Below are the codes to make everything work, most of the methods not shown here can be called manually (they are shown in each individual class' synopsis, if any)
```java
import assets.*;

import java.util.*;

public class Synopsis {
	
	public static void main (String args[]) {
		
		// Set up the map and gps
		CityMap mapA = new CityMap();
		GPS gps = new GPS();
		
		// process and get ready all the paths
		// "List" in java.util and java.awt conflicts
		ArrayList<java.util.List<Station>> partialNetworks =  mapA.processPartialNetwork(true);
		ArrayList<java.util.List<Station>> fullPaths = mapA.processFullNetwork(partialNetworks, 5, false); // don't changed the number 5, allowed to increase it but not too much
	
		// Show GUI after processing the full paths
		Screen GPS_Screen = new Screen();
		GPS_Screen.sendMap(mapA);
		GPS_Screen.renderGraphics(); // only stations connected using CityMap::connectBridges will be drawn
				
		System.out.println("Screen rendering complete!");
		
		// wait for the details in gui to be filled
		// the screen is instance use only, so we need to terminate everything for every search :)
		while (!GPS_Screen.infoIsAssumedComplete) {
			try {
				System.out.println("Screen info is not complete, waiting for another 5 seconds...");
				Thread.sleep(5000);
			} catch (Exception e) {
				break;
			}
		}
		
		System.out.println("Screen info is complete, procedding to the algorithm\n"); // remove if necessary
		
		// get data from the gui
		int constraintData = GPS_Screen.getConstraintData(); // actual capacity or the weight 
		String constraintChoice = GPS_Screen.getConstraintChoice().toLowerCase(); // Either "Capacity" or "Weight"
		String selectedAlgorithm = GPS_Screen.getSelectedAlgorithm(); // Either "CSP" or "COP"
	
		// For COP, choose any station that seem farthest away to get some correct fascinating results
		// getting everything ready
		ArrayList<Station> stationNames = GPS_Screen.getSelectedStations();
		
		ArrayList<java.util.List<Station>> availablePaths;
	
		String start = stationNames.get(0).name;
		String destination = stationNames.get( stationNames.size()-1 ).name;
		availablePaths = mapA.showAvailablePaths(start, destination, fullPaths, false, constraintData, false); // 10 is the capacity
		
		// start routing based on the algorithm chosen
		if (selectedAlgorithm == "CSP") {
			
			Station[] chosenPath_CSP = gps.findChosenPath(availablePaths, stationNames, constraintData);
			
			if (chosenPath_CSP != null) {
				
				// to draw the chosen path on the screen
				GPS_Screen.setChosenPathToHighlight(chosenPath_CSP);
				
				System.out.println("Showing the path to use (CSP)...");
				for (int i=0; i <= chosenPath_CSP.length - 1; i++) {
					System.out.print(chosenPath_CSP[i]);
					if (i != chosenPath_CSP.length - 1) {
						System.out.print(", ");
					}
				}
			} else {
				System.out.println("No path found, make sure your " + constraintChoice +" is big enough");
				System.out.println("And also make sure that you're not returning back to CW or intersecting it");
				System.out.println("Or it's just that the system encountered a loop and died :)");
			}
			System.out.println("\n");
			
		} else if (selectedAlgorithm == "COP") {
	
			Station[] chosenPath_COP = gps.findChosenPath(availablePaths, start, destination);
			
			if (chosenPath_COP != null && constraintData > 0) {
				
				// to draw the chosen path on the screen
				GPS_Screen.setChosenPathToHighlight(chosenPath_COP);
				
				System.out.println("Showing the shortest path(COP):");
				for (int i=0; i<chosenPath_COP.length; i++) {
					System.out.print(chosenPath_COP[i].name);
					if (i != chosenPath_COP.length - 1) {
						System.out.print(", ");
					}
				}
				System.out.println("");	
			} else {
				System.out.println("No path found, make sure your " + constraintChoice +" is big enough");
				System.out.println("And also make sure that you only select the starting and destination stations only");
			}
			System.out.println("\n");
			
		}
		
	}
}
```

## GUI on Different Platforms
### On Windows 10
![GUI under Windows 10](https://github.com/JADE-Cars/JADE-Cars-MapSearching/blob/main/images/gui_windows10.png)

### On Linux Mint 20
![GUI under Linux Mint 20](https://github.com/JADE-Cars/JADE-Cars-MapSearching/blob/main/images/gui_linuxMint20.png)

### How it Works?
- THE GUI IS INSTANCE USE. SO, EVERY TIME A SEARCH IS DONE, YOU'LL NEED TO TERMINATE EVERYTHING AND START AGAIN :)
- The \~\~STATION LIST\~\~ is the list of stations that you can choose. By default, CW is already selected. The selected station names will be displayed on the console.
- After selecting all your desired stations, click "Register Route", this will display all the stations selected according to sequence at the bottom of the screen (with purple background). You can also click the "Register Route" button every time you select a station.
- The "Clear Route" button will remove all selected stations, but the default CW station will remain. CW is assummed to be the company+warehouse. The W of CW has nothing to do with other W* stations :)
- The constraint select box is for you to specify either to use the capacity or weight.
  - At the right side of the constraint box is an input field, make sure that you specify a big number to allow all the possible routes to be processed. **Remember to hit the ENTER button after entering  the value**, it will be disabled after that (the color becomes lighter)
- The algorithm part is for you to select either CSP or COP for the search
- The "Route" button will finally process the information entered and the search starts.
### Do
- Enter every field and chose every option that can be chosen
- Enter an appropriate value (not too big or too small) for the capacity or the weight to allow the searching algorithms to work correctly :)
- Make sure that you hit the ENTER key after entering the value for the constraint (either capacity or weight). The constraint value field will be disabled after the ENTER is hit.
- Click on the screen (anywhere in the pink area) if the original map (with green stations) disappear after the chosen route is highlighted or when the GUI is minimized and maximized again. 

### Don't
- Don't ever ever ever ever press the SPACEBAR right after clicking any buttons. The buttons include "Register Route", "Clear Route", and "Route" buttons.
- Don't try to press the "X" button of the GUI, that part was not implemented on purpose :)

### Other things to take note of
- It is advisable to make sure that the number of stations (green dots) chosen follows the following:
  - For **CSP**, make sure **not to create a loop**. This algorithm should be able to take in any number of stations.
    - The station roads might be bi-directional, but the routing algorithms don't care about that :)
  - For **COP**, make sure you **only** choose the **starting and destination stations**.
  - If it is obvious that your chosen path is correct and the searching algorithm should work, but the system terminates or can't produce any route, try again, it has something to do with the way the algorithms are coded
- Entering a huge value for the constraint might cause the algorithm to go wrong :)
- Not following the above instructions will most likely end up with strange or fascinating routes :) Sometimes the results are still correct
- The two algorithms have some randomness in them.

## Classes & Methods
### 1. assets::Station

#### Synopsis
```java
import assets.Station;
import java.util.*;

...

Station home = new Station(30, 50, "Home");
Station mall = new Station(100, 100, "Mall");
Station airport = new Station(500, 500, "Airport");

home.connectTo(mall, 50);
mall.connectTo(home, 50); // an explicit loop, be careful
mall.connectTo(airport, 25);

// connections are bi-directional by default
System.out.println("Home<->mall: " + home.isConnectedTo(mall));
System.out.println("Mall<->Home: " + mall.isConnectedTo(home));

System.out.println("Airport<->Mall: " + airport.isConnectedTo(mall));
System.out.println("Mall<->Airport: " + mall.isConnectedTo(airport));

// indirect connections can't be processed, see CityMap
System.out.println("Home<->Airoport: " + home.isConnectedTo(airport));

System.out.println("d(Home<->Mall)=" + home.distanceFrom(mall));
System.out.println("d(Mall<->Home)=" + mall.distanceFrom(home));
System.out.println("d(Airport->Mall)=" + airport.distanceFrom(mall));

// indirect connections can't be processed, see CityMap
try {
	System.out.println("d(Home->Airport)=" + home.distanceFrom(airport));
} catch (NullPointerException e) {
	System.out.println("Home is not directly connect to the airport.");
}

System.out.println("");

home.verbose_distanceFrom(mall);
airport.verbose_distanceFrom(mall);

// indirect connections can't be processed, see CityMap
try {
	home.verbose_distanceFrom(airport); // error
} catch (NullPointerException e) {
	System.out.println("Home is not directly connect to the airport.");
}

System.out.println("");

// show the details of each station's connections
home.showDetails();
mall.showDetails();
airport.showDetails();
```

#### Station(int x, int y, String name)
The constructor to create a new *Station*, x and y are the location to be drawn on the screen and have nothing to do with the distance in *CityMap::createBridges*

#### public void connectTo(Station s, int distance)
Connects the current *Station* to another *Station*. This eliminates the need of an edge. *Station* is the vertex :)

Whenever a *Station* is connected to this current station, both *Station*'s connections list will be updated. This works because a *Station* must be passed in and therefore both station's infomation can be obatin in one go

#### public boolean isConnectedTo(Station s)
Checks if this current station is connected to another *Station* s

#### public int distanceFrom(Station s)
The distance from this current station to another *Station* s

#### public void verbose_distanceFrom(Station s)
More text :)

#### public Map<Station, Integer> allConnections()
Returns the whole list of connections from this station to all other stations, if any

This method is called by *showDetails()*

#### public String toString()
This method is overridden to return a string that resembles the Cartesian coordinates as much as possible for readability

#### public void showDetials()
Calls *toString()* and adds on more details


### 2. assets::CityMap

#### Synopsis
```java
import assets.Station;
import asets.CityMap;
import java.util.*;

...

// create maps
CityMap firstMap = new CityMap();
CityMap secondMap = new CityMap();

// add on stations to second map
Station home = new Station(30, 50, "Home");
Station mall = new Station(100, 100, "Mall");
Station airport = new Station(500, 500, "Airport");

home.connectTo(mall, 50);    secondMap.createBridges(home, mall);
mall.connectTo(home, 5);     secondMap.createBridges(mall, home); // this will cause a loop
mall.connectTo(airport, 25); secondMap.createBridges(mall, airport);

// link all the connections
System.out.println("First map's partial network:");
ArrayList<List<Station>> firstMapPartialNetwork = firstMap.processPartialNetwork(true);

System.out.println("");

System.out.println("Second map's partial network:");
ArrayList<List<Station>> secondMapPartialNetwork = secondMap.processPartialNetwork(true);

// the number 5 is the minimum iterations needed to link all the default stations
ArrayList<List<Station>> secondMapFullNetwork 
	= secondMap.processFullNetwork(secondMapPartialNetwork, 5, false);

// capacity of 3 will exclude the starting destination (Home)
ArrayList<List<Station>> pathsFromHomeToAirport = secondMap.showAvailablePaths("Home", "Airport", secondMapFullNetwork, false, 4, false);
System.out.println(pathsFromHomeToAirport); // a lot of redundant paths
```

#### CityMap()
The default constructor. This will set and connect all the Stations. You can also add-on connections if you want to :)

The number of layers to pass to asset::CityMap::processFullNetwork is 5

#### public void createBridges(Station s1, Station s2)
Connects 2 *Stations* in one go. This is the minimum stations that can be connected.

This means that the stations are bi-directional. However, this is not taken into consideration during the route searching process :)

#### private void updateStationRecord(Station s1, Station s2)
This method will be called by the *createBridges* method. The record is needed for the actual searching based on user input.

#### public ArrayList<List\<Station>> processPartialNetwork(boolean verbose)
Stores all the connections created by *CityMap::creteBridges()*. At this stage, we only have all the partial network. There might still be some connections that can/needs to be combined into a single connection. See also *processFullNetwork()*

If **verbose** is set to true, it will show all the partial networks that will go through the linking process.
  
#### public ArrayList<List\<Station>> processFullNetwork(ArrayList<List\<Station>> _partialNetworks, int maxLayers, boolean verbose)
This method calls the private method *linkAllConnectingStations()* to combine the bridges. See *linkAllConnectingStations()* for more details.

*maxLayers* is the number of layers to be proccessed. Draw your map on a piece of papaer first and then determine how many layers it will be :) A branching network will caused an extra layer to exists.

If **verbose** is set to true, it will display the combined-once partial network and then notify if there are any partial networks that can be link for each iteration.

This method will store every path in *_partialNetworks* into a new ArrayList<List<Station>>, any linking partial network will then be processed and added in. This means that there might be some "redundant" partial paths (before linking then together) included which might be useful depending on the use case.

#### private ArrayList<List\<Station>> linkAllConnectingStations(ArrayList<List\<Station>> _partialNetworks, boolean verbose)
This method combines the partial networks by looping through the ArrayList ONCE only and each *List<Station>* within will be checked against another *List<Station>*. Therefore, only 2 linking partial networks will be combined together each time. If there are more than 2 partial networks that should be linked toegther, only the first 2 (that the JVM happen to see) will be processed. This methods returns the same list as the *processFullNetwork()* method. This method can be called as many times as needed.

If **verbose** is set to true, it will show all the paths that meets the requirements ie. paths that include the specified start and destination stations.

*This method should be optimized further in the future (The starting station should be in the first record while the destination station should be in the last record and not somewhere in the middle due to some simple geometry concepts)

#### public ArrayList<List\<Station>> showAvailablePaths(String start_name, String destination_name, ArrayList<List\<Station>> processedPaths, boolean verbose, int capacity, boolean analyseCapacity)
Process the available paths for the specified starting and destination stations. If the number of stations exceed the *capacity*, it will be ignored. 

If **verbose** is set to true, it will display all the available paths. This output is only good for debugging. If you need to generate a nicer display, just override the Station.toString() method or manually change the display in this method.

Setting **analyseCapacity** to true will give details about the ignored paths.

### 3. assets::GPS
#### public Station[] findChosenPath(ArrayList<java.util.List\<Station>> paths_2_filter, String check_start_name, String check_destination_name)
This is the last processing step. This will chose the shortest path for the Delivery Agent. This only checks for the starting and destination stations.

This method utilizes the **COP** concept.

#### public Station[] findChosenPath(ArrayList<java.util.List\<Station>> paths_2_filter, ArrayList\<Station> stationNames, int capacity)
This is the last processing step. Unlike the method above, this method will try to find a path that includes all the stations (according to sequence) specified in *stationNames*. Take note that this method doesn't call the above method eventhough it should :) If the path is invalid or no path can be determined for the specied stations, *null* is returned

This method utilizes the **CSP** concept.

This method was suppose to refine the above method, but it ended up an idependent method :)

*Take note that this algorithm doesn't guarantee correct results all the time, if any paths are situated in a looping network, it is very likely for the algorithm to fail. This has something to do with the way all the assets were coded.

#### private static Station[] schwartzianTransform(ArrayList<java.util.List\<Station>> paths_2_filter)
Schwartzian transformation used by *findChosenPath* (COP) method

#### private static Station[] correctedDirection(Station[] shortestPath, String check_start_name, String check_destination_name)
This method tries to correct the direction of the chosen path if the filtered path is not in the correct order. It's either in the correct order of the other way round.

#### public static void drawMap(Station[] stations)
Draws all the connected stations and the details on the screen. The map might be gone as soon as it leaves the actual computer screen. This can b esolved by clicking the background (pink area) to repaint the map.

### 4. assets::Screen
This class extends the **Frame** class and is used by **GPS**. The graphics is generated using java.awt.
This class should be made into a local variable inside assets::GPS, but anyway

#### Screen ()
This sets the title/name of the program and overrides the following listeners:
 - mousePressed(MouseEvent e)
   - to cause the map to be redrawn if the gui window is minimized and open again
   - sendMap() and repaint() is called here

This method also creates all the necessary components for the GUI
 
#### public void sendMap(CityMap map)
This method will process the iterator of all the values of CityMap's stationDictionary variable and pass the whole map to the class for repainting the screen

#### public void renderGraphics() 
Renders the display. The city map will be drawn.

#### private Panel renderStationListLabel()
Create the panel for the label with the text \~\~STATION LIST\~\~ at the top right corner of the screen

#### private Panel renderStationList()
Creates the panel for the list of station names. This is a selection box

This method will be called by *renderGraphics()*

#### private java.awt.List poolList()
Called by *renderStationList()* mentioned above. This is the data/names of the stations

This method will be called by *renderStationList()* to populate the list.

*Take note that the return type might conflict with java.util.List

#### private Panel renderRegisterRouteButton(Button registerRouteButton, Button clearRouteButton)
Creates the panel for the 2 buttons--register route and clear route.

*registerRouteButton* and *clearRouteButton* correspond to the "Register Route" and "Clear Route" buttons respectivly. This passing of parameters is to make the action listener work.

This method will be called by *renderGraphics()*

CHANGELOG
28 MAY 2021
- Changed the constructor of Station to public
- Add synopses for some of the Station and CityMap classes
- Made minor corrections to README
- Added a RELEASENOTES file to keep track of the releases

26 MAY 2021
- Created the jar file assets v1.0.0

25 MAY 2021
- Corrected the top left corner of the map a little bit, the CSP should work better by now

21 MAY 2021
- Added the benchmarking codes for CSP and COP

20 MAY 2021
- Added more stations and changed some colors
- Tweaked the GUI a little bit
- Added more description to TestScript::main

19 MAY 2021
- Add all the listeners needed for the GUI components.
- The screen can now paint the chosen path determined by the algorithm
- Rearranged the remote comtrol's components :)
- Moved the screen dimension into Screen, so no more setting the screen size manually
- Fixed the CSP algorithm, it is now showing the correct path (most likely)
- Updated the TestScript
- Changed the details of company station to cw with the name of CW

18 MAY 2021
- Modified the layout of the GUI a bit and added some more components
- Added the logic ie. event listeners for most of the components (buttons, text field, list)

17 MAY 2021
- Overloaded GPS::findChosenPath. The original version is used to check ONLY the starting station and the destination station. 
  The newer version is used to look for the path that includes the extra stations in between.
  *Take note: GPS::findChosenPath WITH capacity utilizes the CSP concept, will return null if the selected path is not valid
              GPS::findChosenPath WITHOUT capacity utilizes the COP concept 
- Modified the display of the map and relocated quite some stations
- Added more stuff to the gui (remote control)
- CityMap::createBridges for 3 stations in one go is not longer available due to some unexpected bridge issues that might occur
- Due to new stations and connections added, CityMap::processFullNetwork's number of layers should be set to 5
- TestScript is now more readable :)

7 MAY 2021
- CityMap::showAvailablePaths now supports checking against capacity
- Changed Station::toString's return value to resemble the Cartesian coordinates
- Some more stations were connected in the map by default
- Screen now displays the distance between stations

5 MAY 2021
- This was forked into JADE-Cars as some codes in the original MapSearching has been edited to suite the whole project in Eclipse
- On packaging the codes:
  1. Station, CityMap, GPS and Screen were packaged under "assets" namespace.
  2. TestScript.java was left out to be the general script to run and test all the assets
- Solved the problem with screen going blank after reopening the minimized gui window(if it appears blank, click the screen)
- Moved the Screen class into another file, Eclipse was complaining about that :)
  - Added MouseListener and MouseAdapter to Screen
- Screen::sendStationList was redefined as Screen::sendMap()
- Successfully intergrated CityMap and GPS into MasterRoutingAgent in setup()

1 MAY 2021
- Refactored the codes in GPS::findChosenPath with the following sub-methods:
  1. GPS::correctedDirection
  2. GPS::schwartzianTransform
- Added a route direction checking feature (only applicable for GPS::findChosenPath) due to the way paths are filtered
- The GPS screen is finally displaying the stations and the roads
- Coordinates of Stations in CityMap were changed to some values that makes sense :)


29 APRIL 2021
- Updated the followings:
  1. CityMap.java
  2. GPS.java
  3. TestScript.java
- Updated the documentation (README)


28 APRIL 2021 and beyond :)
- Created some of the classes
- Updated Station.java and some other files
- Forked the proginal codes into Jade-Cars' repo


assets/Screen.java
package assets;

import java.awt.*;
import java.awt.event.*;
import java.util.*;

public class Screen extends Frame {
	//private Station dummy = new Station(300, 300, "Dummy");
	private Iterator<Station> stations_it = null;
	private CityMap localMap = null;
	private Station[] chosenPath;
	private boolean hasChosenPath = false;
	
	private int stationDiameter = 20;
	private int stationRadius = stationDiameter / 2;
	
	private java.awt.List stationList = new java.awt.List();
	private Button routeButton = new Button("Route");
	private TextField routeTF = new TextField();
	private ArrayList<String> selectedStationNames = new ArrayList<String>();
	private Button registerRouteButton = new Button("Register Route");
	private Button clearRouteButton = new Button("Clear route");
	private TextField constraintTF = new TextField();
	private Choice constraint = new Choice();
	private Choice algorithmMenu = new Choice();
	
	private int constraintData;
	private String constraintChoice;
	private String selectedAlgorithm;
	
	public boolean infoIsAssumedComplete = false; // out there, make an infinity loop to check this
	
	public Screen() {
		
		super("COS30018 Vehicle Routing System - GPS");
		// listeners		
		this.addMouseListener( new MouseAdapter() {
			public void mousePressed(MouseEvent e) {
				Screen gps_screen = (Screen) e.getSource();
				gps_screen.sendMap(localMap);
				gps_screen.repaint();
			}
		} );
		
		// add listeners
		routeButton.addActionListener( new ButtonListener() );
		registerRouteButton.addActionListener( new ButtonListener() );
		clearRouteButton.addActionListener( new ButtonListener() );
		// stationList.addItemListener( new ListListener() ); // this must go into renderStationList()
		selectedStationNames.add("CW");
		
		constraintTF.addActionListener( new ConstraintTFActionListener() ); // disable it by hitting enter key
		constraint.addItemListener( new ConstraintItemListener() );
		algorithmMenu.addItemListener( new SelectedAlgorithmItemListener() );
	}
	
	public void sendMap(CityMap map) {
		this.stations_it = map.stationDictionary.values().iterator();
		this.localMap = map;
	}
	
	// the actual GUI ie the Screen
	public void renderGraphics() {
		setBounds(300, 100, 688, 530);
		setBackground(Color.PINK);
		setLayout(null);
		//setResizable(false);

		Panel stationListLabelPanel = renderStationListLabel();
		Panel stationListPanel = renderStationList();
		Panel details = renderDetails(constraintTF, constraint, algorithmMenu);
		Panel routeDisplay = renderRouteDisplay(this.routeTF);
		Panel routeButton = renderRouteButton(this.routeButton);
		Panel registerRoute = renderRegisterRouteButton(this.registerRouteButton, this.clearRouteButton);
		
		add(stationListLabelPanel); add(stationListPanel); add(details); add(routeDisplay); add(routeButton); add(registerRoute);
		
		setVisible(true);
	}
	
	private Panel renderStationListLabel() {
		Panel stationListPanel = new Panel(); 
		stationListPanel.setBounds(500, 30, 180, 30);
		stationListPanel.setLayout(new GridLayout(1, 1));
		stationListPanel.setVisible(true);
		stationListPanel.setBackground(new Color(220, 220, 220));
		
		Label stationListLabel = new Label("~ ~ STATION LIST ~ ~");
		stationListLabel.setAlignment(Label.CENTER);
		stationListPanel.add(stationListLabel);
		
		return stationListPanel;
	}
	
	private Panel renderStationList() {
		Panel stationListPanel = new Panel();
		stationListPanel.setBounds(500, 60, 180, 280);
		stationListPanel.setLayout(new GridLayout(1, 1));
		stationListPanel.setVisible(true);
		stationListPanel.setBackground(Color.BLUE);
		
		this.stationList = poolList();
		this.stationList.setMultipleMode(true);
		this.stationList.addItemListener( new ListListener() );
		stationListPanel.add(this.stationList);
		
		return stationListPanel;
	}
	
	private java.awt.List poolList() {
		
		java.awt.List stations = new java.awt.List();
		ArrayList<String> stationNames2sort = new ArrayList<String>();;
		
		Iterator<String> it =  localMap.stationDictionary.keySet().iterator();
		String stationName;
		for (int i=0; it.hasNext(); i++) {
			stationName = it.next();
			if (stationName == "CW") {
				continue;
			}
			stationNames2sort.add( stationName );
		}
		Collections.sort(stationNames2sort);
		
		for (int i=0; i <= stationNames2sort.size()-1; i++) {
			stations.add(stationNames2sort.get(i));
		}
		
		return stations;
	}
	
	private Panel renderRegisterRouteButton(Button registerRouteButton, Button clearRouteButton) {
		Panel registerRouteButtonPanel = new Panel();
		registerRouteButtonPanel.setBounds(500, 340, 180, 50);
		registerRouteButtonPanel.setLayout(new GridLayout(1, 2));
		registerRouteButtonPanel.setVisible(true);
		registerRouteButtonPanel.setBackground(Color.GREEN);
		
		registerRouteButton.setActionCommand("register route");
		clearRouteButton.setActionCommand("clear route");
		
		registerRouteButton.setBackground(new Color(210, 210, 210));
		
		registerRouteButtonPanel.add(registerRouteButton);
		registerRouteButtonPanel.add(clearRouteButton);
		
		return registerRouteButtonPanel;
	}
	
	private Panel renderDetails(TextField constraintTF, Choice constraint, Choice algorithmMenu) {
		Panel details = new Panel(); 
		details.setBounds(500, 390, 180, 60);
		details.setLayout(new GridLayout(2, 2));
		details.setVisible(true);
		details.setBackground(Color.CYAN);

		constraint.add("-Constraint-");
		constraint.add("Capacity");
		constraint.add("Weight");
		// constraint.setSize(60, 30); // doesn't work
		
		Label algorithm = new Label();
		algorithm.setText("Algorithm: ");
		
		algorithmMenu.add("-Algorithm-");
		algorithmMenu.add("CSP");
		algorithmMenu.add("COP");
		
		details.add(constraint); details.add(constraintTF);
		details.add(algorithm); details.add(algorithmMenu);
		
		return details;
	}
	
	private Panel renderRouteDisplay(TextField routeTF) {
		Panel routeDisplay = new Panel();
		routeDisplay.setBounds(8, 500, 680, 25);
		routeDisplay.setLayout(new GridLayout(1, 1));
		routeDisplay.setVisible(true);
		routeDisplay.setBackground(Color.YELLOW);
		
		// TextField
		routeTF.setBackground(new Color(250, 150, 250));
		//routeTF.setEnabled(false);
		routeTF.setText("Select the stations on the top right panel and click \"Register Route\"");
		
		routeDisplay.add(routeTF);
		
		return routeDisplay;
	}
	
	private Panel renderRouteButton(Button routeButton) {
		
		Panel routeButtonPanel = new Panel();
		routeButtonPanel.setBounds(500, 450, 180, 50);
		routeButtonPanel.setLayout(new BorderLayout());
		routeButtonPanel.setVisible(true);
		routeButtonPanel.setBackground(Color.ORANGE);
		
		routeButton.setActionCommand("route");
		routeButton.setBackground(new Color(220, 220, 220));
		routeButtonPanel.add(routeButton, BorderLayout.CENTER);
		
		return routeButtonPanel;
	}
	
	public void setChosenPathToHighlight(Station[] chosenPath) {
		this.chosenPath = chosenPath;
		hasChosenPath = true;
		repaint();
	}
	
	public int getConstraintData() {
		return constraintData;
	}
	
	public String getConstraintChoice() {
		return constraintChoice;
	}
	
	public String getSelectedAlgorithm() {
		return selectedAlgorithm;
	}
	
	public ArrayList<Station> getSelectedStations() {
		
		ArrayList<Station> selectedStations = new ArrayList<Station>();
		
		Station station;
		for (int i=0; i < selectedStationNames.size(); i++) {
			station = localMap.stationDictionary.get( selectedStationNames.get(i) );
			selectedStations.add( station );
		}
		
		return selectedStations; 
	}
	
	public void paint(Graphics g) {
		Color c = g.getColor();
		
		Station station = null;
		while (stations_it.hasNext()) {
			station = stations_it.next();

			// circles
			g.setColor(new Color(110, 200, 90));
			g.fillOval(station.x, station.y, stationDiameter, stationDiameter);
			
			// station names
			g.setColor(Color.BLACK);
			g.drawString(station.name, station.x + 20, station.y + 10);
			
			// draw roads
			Iterator<Station> connected_it  = station.allConnections().keySet().iterator();
			while (connected_it.hasNext()) {
				Station connectedStation = connected_it.next();
				
				// road
				g.setColor(Color.WHITE);
				g.drawLine(station.x + stationRadius, station.y +stationRadius, 	
							connectedStation.x +stationRadius , connectedStation.y +stationRadius);
				
				// distance
				g.setColor( new Color(121, 9, 143) );
				g.drawString(String.valueOf(station.distanceFrom(connectedStation)), 
						(station.x + connectedStation.x)/2, (station.y + connectedStation.y)/2);
			}
		}
		
		// paint chosenPath if any
		if (hasChosenPath) {
			//System.out.println("Painting extra stuff");
			
			// circles
			for (int i=0; i < chosenPath.length; i++) {
				// chosen circles
				g.setColor(new Color(144, 78, 230));
				g.fillOval(chosenPath[i].x, chosenPath[i].y, stationDiameter, stationDiameter);
			}
			
			for (int i=1; i < chosenPath.length; i++) {
				// road
				g.setColor(new Color(144, 78, 230));
				g.drawLine(chosenPath[i-1].x + stationRadius, chosenPath[i-1].y +stationRadius, 	
							chosenPath[i].x +stationRadius , chosenPath[i].y +stationRadius);
			}
		}
		
		// to test repaint() logic
		/*
		g.setColor(Color.WHITE);
		g.fillOval(dummy.x, (int)(Math.random()*500 + 1), stationDiameter, stationDiameter);
		g.drawString(dummy.name, dummy.x + 20, dummy.y + 10);
		*/
		
		g.setColor(c);		
	}
	
	private class ButtonListener implements ActionListener {
		
		public void actionPerformed(ActionEvent e) {
			//System.out.println("Button pressed!");
			Button someButton = (Button) e.getSource();
			
			// we might still have some more buttons
			if (someButton.getActionCommand() == "route") {
				
				System.out.println("Route button pressed");
				infoIsAssumedComplete = true;
				
			} else if (someButton.getActionCommand() == "register route") {
				
				System.out.println("Register route button pressed");
				System.out.println("Selected stations: " + selectedStationNames);
				
				routeTF.setText(selectedStationNames.toString());
				
			} else if (someButton.getActionCommand() == "clear route") {
				
				System.out.println("Route cleared");
				
				selectedStationNames = new ArrayList<String>();
				selectedStationNames.add("CW");
				routeTF.setText(selectedStationNames.toString());
				
				System.out.println("Selected stations: " + selectedStationNames);
				
			}
		}
	}
	
	
	private class ListListener implements ItemListener {
		public void itemStateChanged(ItemEvent e) {
			java.awt.List selection = (java.awt.List) e.getSource();
			System.out.println(selection.getSelectedItem() + " was selected");
			
			if (selection.getSelectedItem() != null) {
				selectedStationNames.add(selection.getSelectedItem());
			}
			
			selection.deselect(selection.getSelectedIndex());
		}
	}
	
	private class ConstraintItemListener implements ItemListener {
		public void itemStateChanged(ItemEvent e) {
			Choice constraint = (Choice) e.getSource();
			constraintChoice = constraint.getSelectedItem();
			System.out.println("Constraint " + constraintChoice + " chosen");
		}
	}
	
	private class ConstraintTFActionListener implements ActionListener {
		public void actionPerformed(ActionEvent e) {
			//System.out.println("Enter key pressed for constraint text field!");
			TextField constraintTF = (TextField) e.getSource();
			constraintTF.setEnabled(false);
			constraintData = Integer.parseInt( constraintTF.getText() );
			System.out.println("Constraint data: " + constraintData);
		}
	}
	
	
	private class SelectedAlgorithmItemListener implements ItemListener {
		public void itemStateChanged(ItemEvent e) {
			Choice algorithm = (Choice) e.getSource();
			selectedAlgorithm = algorithm.getSelectedItem();
			System.out.println("Algorithm selected: " + selectedAlgorithm);
		}
	}
}